### ISSUE TITLE
Clock synchronisation between Sensors/Things

### ISSUE THREAD

**Issue by abourdon** [opengeospatial/sensorthings#47]
Hi,

We have a use case that uses more than 2 `Sensor`s that send `Observation`s about **the same `ObservedProperty`**. To have a global view of all `Observation`s about this same `ObservedProperty`, we then have to consolidate all `Sensor`'s `Observation`s.

But to correctly do that, we have to make sure that all `Sensor`s (or associated `Thing`s) share the same clock. This way, we will be able to sort `Observation`s against their `phenomenonTime`s or `resultTime`s. Otherwise, it would be difficult to have a temporal view of our data.

A possible solution to our problem would be to have a _reference time_ for each `Observation`. Thus, whatever the time offset between multiple `Sensor`s `Observations`, we will be able to synchronize them with this _reference time_.

A simple implementation of this solution could be to add a `createdAt` (or name it as you want) `Observation`'s **property** that will be set before any `Observation` is sent to our SensorThings service. This way, we will need to have an ad-hoc solution (any solution could be imagined) that set this _reference time_ anytime we send an `Observation`.

However, this implementation needs to have this ad-hoc solution, and so anytime someone will face this issue, he will need to have this additional layer, that could be a drag on the use of the SensorThings API.

A better implementation could be to have this `createdAt` value as an `Observation`'s **attribute** (so handled by the SensorThins API model and no more as an implementation-specific property). This way, this value could be set either:
- by any ad-hoc solution
- or by the SensorThings service itself when this value is missing, by using its current time (default behavior).

What do you think about it? Maybe I'm totally wrong and there is another way to do it, that's why I will be happy to know your opinion on it :-)

Regards,
Aurélien

**Comment by mjacoby** [comment]
To my understanding, clock-synchronization is way beyond what the standard covers and probably should cover.

However, there might be a simple solution to your problem. According to the standard, Observation has a property `phenomenonTime` that describes a time instant or period when the observation happens. If it is not set when creating an observation, the server assigns the current server time. (see description in table 18 here http://docs.opengeospatial.org/is/15-078r6/15-078r6.html#31)

So basically this is exactly what you propose with the `createdAt` property, isn't it?

**Comment by abourdon** [comment]
@mjacoby, yes but only using `phenomenonTime` is not sufficient. There could be `Sensor` that already set this value but not all `Sensor`s. By using a new attribute, we can address both cases. 

I would like to have a more general solution to address this issue whatever the possibilities of a `Sensor`.

**Comment by liangsteve** [comment]
@abourdon Thanks for providing the background of the potential issue.  Can you please write a precise description or a definition of this attribute you propose?  It will help the SWG to understand the proposed attribute better.  Thanks.

**Comment by abourdon** [comment]
@liangsteve To sum up my proposal I would say:

Create a new `Observation`'s attribute, named `createdAt`, that corresponds to the time this `Observation` has been created from the SensorThings server side. This value must be set by the SensorThings server itself by using the current SensorThings server's time.

Thinking about this further, I finally don't think it is necessary for a client to override it. This way, value will not be provided by client but always generated by the SensorThings server and available when requesting `Observation`s.

This way, it could solve my issue without being too specific about it (and be more in phase with standard vision, @mjacoby).

What do you think?

**Comment by hylkevds** [comment]
What would happen to this time instant when the entity is updated?

**Comment by abourdon** [comment]
@hylkevds, in this case, the `createdAt` attribute will remain unchanged. We could think about a secondary attribute, `updatedAt` or `lastUpdatedAt` (for instance) that will contain the last time entity has been updated.

And as we talk about the term _entity_, we could also generalize these `createdAt` and `lastUpdatedAt` attributes to be part of all entites, as they are not only restricted to `Observation`s.

**Comment by mjacoby** [comment]
@abourdon I get your point and without deeper analysis adding the property `createdAt` and `updatedAt` looks like a viable option.
The only real question to me is whether this is such an essential functionality that it should be included into the standard or we try to keep the standard as simple and minimalistic as possible...
There are lots of viable application scenarios where these properties are not needed and in this cases we "pollute" each response to a GET of an entity with two additional fields. 
If we keep adding additional properties to an entity/entities in the future, this will become more and more of a problem, i.e. slowing down performance and making the standard complex.

Therefore, I would propose to change the standard to explicitly focus on extensibility. This could be done either by allowing additional fields in JSON instead of failing (what is currently the defined behaviour) or by adding an attributes `properties` of a list of key-value-pairs to all entities /like our FROST Server implementation does). 

This way it would be possible to easily adjust an existing server implementation, for example to provide the additional properties you suggested.

**Comment by abourdon** [comment]
@mjacoby I totally agree to keep the **core** model as simple as possible.

And to go further with your idea of extension, we could imagine to define a set of **optional** model properties that a server implementation **could** allow following a specific request option or specific server configuration. If an **optional** model property make sense to be available whatever the server implementation, we could include it to the SensorThings API specification to restrict any server implementation to follow the same naming convention.

As a direct application, we could imagine our two `createdAt` and `updatedAt` as an optional model properties that could be activated when the a specific request option or a server configuration is set. As these optional model properties are generic enough and very close to what the definition of _persisted_ entity is, we could imagine to add them to the SensorThings API specification.

For instance, based on FROST-Server configuration, it could be defined like this:

```xml
<context-param>
  <description>Display additional entity's timestamps as creation and update time instants. </description>
  <param-name>entity.EnableTimestamps</param-name>
  <param-value>true</param-value>
</context-param>
```

This way, any time we get a SensorThings's entity, we will get its associated `createdAt` and `updatedAt` attributes (defined respectively as `@iot.createdAt` and `@iot.updatedAt` properties for instance):

```json
{
  "@iot.id": 1,
  "@iot.selfLink": "http://example.org/v1.0/Things(1)",
  "@iot.createdAt": "2018-03-17T07:00:00+01:00",
  "@iot.updatedAt": "2018-04-19T08:00:00+01:00",
  "Locations@iot.navigationLink": "Things(1)/Locations",
  "Datastreams@iot.navigationLink": "Things(1)/Datastreams",
  "HistoricalLocations@iot.navigationLink":
  "Things(1)/HistoricalLocations",
  "name": "Oven",
  "description": "This thing is an oven.",
  "properties": {
    "owner": "Noah Liang",
    "color": "Black"
  }
}
```

What do you think?

**Comment by mjacoby** [comment]
That would be one possible way of doing it.
However, it would be quite unflexible as this would only allow to add properties that are explicitly mentioned in the standard. I'm rather envisioning having something like the OData meta data document, i.e. a file that is describing the information mode used in a concrete server instance. This way it would be more open. Additionally, there could be something like well-known models which could either be part of the standard or some best practice document that is more is to extent.

However, these are quite ground-breaking ideas that would introduce lots of other problems. Therefore my current proposal for the standard would be to just allow additional properties in JSON for any entity and to write down that a server should rather ignore unknown properties instead of failing. This is very straight forward to implement and opens up the standard to be easily extended to application needs.

**Comment by liangsteve** [comment]
One approach is to write an OGC Engineering Report or even better, a Best Practice (with conformance classes), to define a standard way to use it.

**Comment by abourdon** [comment]
@mjacoby maybe I was not clear about my thoughts. I propose to have 2 types of optional properties:
- One type which will be **standard** optional properties, defined in the SensorThings API. This way, we could define most common optional properties with a standardized name where any implementation server could (but not must) handle. This would be the case for `createdAt` and `updatedAt` for instance.
- A second type of optional properties which are **not standard** and uniquely handled by a specific server implementation. Those optional properties are thus not included in the SensorThings API standard and only taking into account by a specific server implementation.

All of these properties can be activated following a dedicated configuration (from a configuration file or request option or whatever).

**Comment by mjacoby** [comment]
Ok, so we are basically on the same side about the functionality.
Open questions are
- Where to define those well-known properties? Options are in the standard or in some best practice document
- How to know that a server supports a well-known property? Could be done through some meta data document as in OData. This is related to another issue I've created some time ago #4

**Comment by KathiSchleidt** [comment]
Short question for clarification - how does createdAt differ from the resultTime?

**Comment by liangsteve** [comment]
7.2.2.3 of OGC 10-004r3 Observation and Measurement - `The attribute resultTime:TM_Instant shall describe the time when the result became available, typically when the procedure (7.2.2.10) associated with the observation was completed. For some observations, this is identical to the phenomenonTime. However, there are important cases where they differ.`

**Comment by liangsteve** [comment]
@KathiSchleidt STA and SOS follow O&M.  As a result, resultTime SHALL follow the definition of mentioned above.

The time an Observation is made publicly available (assuming via STA?) means differently from resultTime as defined in OGC 10-004r3.  As a result, The time an Observation is made publicly available (assuming via STA?) should be using a different attribute.

**Comment by KathiSchleidt** [comment]
so createAt could be understood as the digital object lifespan - would make sense to add
while we're on the topic of observation times, what about times of validation steps? I know that values go through different validation levels at Umweltbundesamt before they're finally accepted as the governmental truth. However, this changes the digital object, so we'd need something the entire set of data curation times, maybe look at 19115?
